### PR TITLE
Add validity check for [Driver][IPAddress] else use docker-machine ip…

### PR DESCRIPTION
**Pull request by gvashchenkolineate**
Think this check should be included in Ansible core too

### CODE REVIEW

**Comment by ximon18** [comment]
Hi @gvashchenkolineate,

Thank you very much for your contribution, it's really appreciated!

I'm a little confused by a couple of things however:

> Could you please tell why didn't you create a PR to original repo and furthermore a PR to include this in Ansible Core?

Firstly, did you see this in README.md?

> This repository is now obsolete
> This plugin was incorported into the official Ansible 2.9 release. See:
> https://docs.ansible.com/ansible/2.9/plugins/inventory/docker_machine.html
> ...
> This plugin has been submitted for inclusion in Ansible core. See [PR #54946](https://github.com/ansible/ansible/pull/54946).

The PR was merged into the Ansible devel branch and was to be included in the Ansible 2.9 release. 

> Think this check should be included in Ansible core too

Secondly, exactly what problem are you trying to solve? I ask because I'm wondering under which circumstances the `docker-machine inspect` command can fail to incude the machine IP address in the output, and yet at the same time the `docker-machine ip` command *does* yield an IP address for the machine?

Thanks,

Ximon

**Comment by gvashchenkolineate** [comment]
Hello @ximon18 !

A little misunderstanding from my side of how a plugin already incorporated into Ansible core can evolve)

I encountered an issue (like [this](https://github.com/moby/moby/issues/35750)) of blank `ansible_host` when using the plugin for a GCP instance created with docker-machine. 
That is `Driver.IPAddress` is blank in `docker inspect`.
And a solution comes for me as [this commit](https://github.com/benroose/ansible-docker-machine-inventory-plugin/commit/3c7d0ae90afc5ae91c436818f489cbc02f941ecd) from @benroose

So I suggested it would help someone else if it will be in Ansible core too.

_Details:
Ansible 2.9.7
Python 2.7.17
Docker-Machine 0.16.2_

Thanks,

Georgy

**Comment by ximon18** [comment]
Thanks. Okay, we have a few things to resolve.

> I encountered an issue (like this) of blank ansible_host when using the plugin for a GCP instance created with docker-machine.

The issue you link to is with Docker itself and concerns a Docker container, while `docker-machine inspect` concerns machines, not containers, so I would not expect the two issues to be related. There does appear to however be some issue in your case with Docker Machine compatibility with GCP.

If this patch worked for you and for @benroose then I'm happy to accept it into my repository, I just wanted to understand why.

I think it would be good to add a comment into the code as to why invoking `docker-machine ip` is necessary, otherwise later we will look back and wonder why we made that change (though perhaps digging through Git history will make it clear, but still I prefer the code to be self evident whereever possible).

Which leads me to the next question. Where is the master repo now? Although [PR #54946](https://github.com/ansible/ansible/pull/54946) it was merged and said it would be included in Ansible 2.9, and although I see the plugin and my name in [the Ansible 2.9 docs](https://docs.ansible.com/ansible/2.9/plugins/inventory/docker_machine.html), I do NOT see my plugin in the Ansible 2.9 source code nor is it still in the devel branch where it went after being merged:

> bcoca merged commit [2cca917](https://github.com/ansible/ansible/commit/2cca9176a7d28d6b4ecd6719f0e2e88f3d0512ed) into ansible:devel on Aug 8, 2019
> 1 check passed
> Shippable Run 136094 status is SUCCESS.

At the time it was "shippable", but now commit [2cca917](https://github.com/ansible/ansible/commit/2cca9176a7d28d6b4ecd6719f0e2e88f3d0512ed) has a red cross which when hovered over indicates that it is no longer shippable.

Additionally the [Ansible 2.9 project board](https://github.com/ansible/ansible/projects/34) is still incomplete and has no updates since late 2019, I wonder what is going on?

Perhaps @felixfontein (sponsor of my work) and @bcoca (reviewer who merged my work into Ansible devel) have some idea? I'm not sure if mentioning them here is enough to get their attention, or if we need to raise this with the Ansible development team via a new issue in Ansible GitHub.

File: docker_machine.py [hunk 1]
```diff
@@ -187,13 +195,20 @@ def _populate(self):
                 env_var_tuples = self._get_docker_daemon_variables(id)
                 if self._should_skip_host(id, env_var_tuples):
                     continue
+                
+                # check for valid ip address from inspect output, else explicitly use ip command to find host ip address
```
Comment (ximon18) [inline docker_machine.py:199]: Please note here that this issue has been seen on GCP.

**Comment by felixfontein** [comment]
@ximon18 the plugin was moved to the [community.general collection](https://github.com/ansible-collections/community.general/), like all docker modules/plugins and a lot other things. (See [here](https://github.com/ansible-collections/overview/blob/master/README.rst) for more information on the move.)

The plugin is also in the Ansible 2.9 code: https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/plugins/inventory/docker_machine.py

The current devel branch is what will be Ansible-base 2.10 "soon". (The last state of the devel branch which contained everything before it was moved to collections is https://github.com/ansible/ansible/tree/pre-ansible-base/)

(Also, the red cross next to the commit is a common sight - after a PR is merged, all tests are re-run for the commit added to `devel`, which often failed because of some random errors in the integration tests. Finally, the 2.9 project board was no longer updated after 2.9.0 was released. The current one is the 2.10 project board: https://github.com/ansible/ansible/projects/39)

**Comment by felixfontein** [comment]
I almost forget, link to the plugin in community.general: https://github.com/ansible-collections/community.general/blob/master/plugins/inventory/docker_machine.py
And if you're interested, docs on how to hack on community.general: https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#contributing-to-collections

**Comment by ximon18** [comment]
Thankyou @felixfontein!

I assume that the correct way to submit a PR now is via the Ansible GitHub repo, is that correct?

Thanks again!

Ximon

**Comment by felixfontein** [comment]
@ximon18 the correct way is to first make a PR for the community.general repo, get it merged, and then (in case of bugfixes) create a backport for the stable-2.9 branch in ansible/ansible.

**Comment by ximon18** [comment]
Thanks @felixfontein !

**Comment by ximon18** [comment]
To speed this along as it seems technically fine I'll merge this into my repo, make a couple of minor changes, then see if I can create a PR against the Ansible repo from my repo. I'll preserve @benroose as the committer of the main change so he gets the credit for that.

**Comment by gvashchenkolineate** [comment]
Thanks everybody!

**Comment by ximon18** [comment]
Thanks all. I have raised https://github.com/ansible-collections/community.general/pull/412.

**Comment by ximon18** [comment]
> The current one is the 2.10 project board: https://github.com/ansible/ansible/projects/39)
@felixfontein: Shouldn't there be a 2.10 roadmap link here? https://docs.ansible.com/ansible/latest/roadmap/index.html 
Perhaps pointing to https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_10.rst?

**Comment by felixfontein** [comment]
@ximon18 it's right now here: https://docs.ansible.com/ansible/devel/roadmap/index.html

**Pull request merged**